src/ethereum/london/spec.py

```diff
@@ -0,0 +1,1136 @@
+    elif parent_gas_used == parent_gas_target:
+        expected_base_fee_per_gas = parent_base_fee_per_gas
+    elif parent_gas_used > parent_gas_target:
+        gas_used_delta = parent_gas_used - parent_gas_target
+        base_fee_per_gas_delta = max(
+            parent_base_fee_per_gas
+            * gas_used_delta
+            // parent_gas_target
+            // BASE_FEE_MAX_CHANGE_DENOMINATOR,
+            1,
+        )
+        expected_base_fee_per_gas = (
+            parent_base_fee_per_gas + base_fee_per_gas_delta
+        )
+    else:
+        gas_used_delta = parent_gas_target - parent_gas_used
+        base_fee_per_gas_delta = (
+            parent_base_fee_per_gas
+            * gas_used_delta
+            // parent_gas_target
+            // BASE_FEE_MAX_CHANGE_DENOMINATOR
+        )
+        expected_base_fee_per_gas = (
+            parent_base_fee_per_gas - base_fee_per_gas_delta
+        )
+
+    ensure(expected_base_fee_per_gas == header.base_fee_per_gas, InvalidBlock)
+
+    parent_has_ommers = parent_header.ommers_hash != EMPTY_OMMER_HASH
+    ensure(header.timestamp > parent_header.timestamp, InvalidBlock)
+    ensure(header.number == parent_header.number + 1, InvalidBlock)
+    ensure(
+        check_gas_limit(header.gas_limit, parent_header.gas_limit),
+        InvalidBlock,
+    )
+    ensure(len(header.extra_data) <= 32, InvalidBlock)
+
+    block_difficulty = calculate_block_difficulty(
+        header.number,
+        header.timestamp,
+        parent_header.timestamp,
+        parent_header.difficulty,
+        parent_has_ommers,
+    )
+    ensure(header.difficulty == block_difficulty, InvalidBlock)
+
+    block_parent_hash = keccak256(rlp.encode(parent_header))
+    ensure(header.parent_hash == block_parent_hash, InvalidBlock)
+
+    validate_proof_of_work(header)
+
+
+def generate_header_hash_for_pow(header: Header) -> Hash32:
+    """
+    Generate rlp hash of the header which is to be used for Proof-of-Work
+    verification.
+
+    In other words, the PoW artefacts `mix_digest` and `nonce` are ignored
+    while calculating this hash.
+
+    A particular PoW is valid for a single hash, that hash is computed by
+    this function. The `nonce` and `mix_digest` are omitted from this hash
+    because they are being changed by miners in their search for a sufficient
+    proof-of-work.
+
+    Parameters
+    ----------
+    header :
+        The header object for which the hash is to be generated.
+
+    Returns
+    -------
+    hash : `Hash32`
+        The PoW valid rlp hash of the passed in header.
+    """
+    header_data_without_pow_artefacts = [
+        header.parent_hash,
+        header.ommers_hash,
+        header.coinbase,
+        header.state_root,
+        header.transactions_root,
+        header.receipt_root,
+        header.bloom,
+        header.difficulty,
+        header.number,
+        header.gas_limit,
+        header.gas_used,
+        header.timestamp,
+        header.extra_data,
```

Comment:
`header.base_fee_per_gas` is missing here.

Reply:
Updated